src/main/java/com/exactpro/th2/crawler/Crawler.java

```diff
@@ -261,166 +290,227 @@ private SendingReport sendEvents(CrawlerId crawlerId, DataProcessorInfo dataProc
                 }
             }
 
-            if (response.hasId()) {
-                RecoveryState oldState = interval.getRecoveryState();
+            if (!response.getIdsMap().isEmpty()) {
+                Map<String, MessageID> responseIds = response.getIdsMap();
+                RecoveryState oldState = RecoveryState.getStateFromJson(interval.getRecoveryState());
 
-                RecoveryState.InnerMessage message = null;
-                String alias = null;
+                Map<CrawlerUtils.AliasAndDirection, RecoveryState.InnerMessage> recoveryStateMessages;
 
-                for (MessageData messageData : messages) {
-                    if (messageData.getMessageId().equals(response.getId())) {
-                        String id = messageData.getMessageId().toString();
-                        Instant timestamp = Instant.ofEpochSecond(messageData.getTimestamp().getSeconds(), messageData.getTimestamp().getNanos());
-                        Direction direction = Direction.valueOf(messageData.getDirection().toString());
-                        long sequence = messageData.getMessageId().getSequence();
-                        alias = messageData.getMessageId().getConnectionId().getSessionAlias();
+                long processedMessagesCount = 0L;
 
-                        message = new RecoveryState.InnerMessage(id, timestamp, direction, sequence);
-                        break;
-                    }
+                for (MessageID messageID : responseIds.values()) {
+                    processedMessagesCount += messages.stream().takeWhile(eventData -> eventData.getMessageId().equals(messageID)).count();
                 }
 
-                if (message != null) {
-                    Map<String, RecoveryState.InnerMessage> lastProcessedMessages = oldState.getLastProcessedMessages();
+                numberOfMessages += processedMessagesCount + diff;
 
-                    lastProcessedMessages.put(alias,
-                            new RecoveryState.InnerMessage(message.getId(),
-                                    Instant.ofEpochSecond(message.getTimestamp().getEpochSecond()),
-                                    Direction.valueOf(message.getDirection().getLabel()),
-                                    message.getSequence()));
+                diff = batchSize - processedMessagesCount;
 
-                    RecoveryState newState = new RecoveryState(oldState.getLastProcessedEvent(), lastProcessedMessages,
-                            oldState.getLastNumberOfEvents(),
-                            numberOfMessages);
+                recoveryStateMessages = messages.stream()
+                        .map(messageData -> {
+                            String alias = messageData.getMessageId().getConnectionId().getSessionAlias();
+                            Instant timestamp = Instant.ofEpochSecond(messageData.getTimestamp().getSeconds(), messageData.getTimestamp().getNanos());
```

Comment:
Don't we have a utility function to convert Timestamp to Instant? If we don't please create one